src/hotspot/share/prims/jvmtiAgent.cpp

```diff
@@ -0,0 +1,620 @@
+  }
+  if (library == nullptr && os::dll_build_name(&buffer[0], sizeof buffer, name)) {
+    // Try the library path directory.
+    library = os::dll_load(&buffer[0], &ebuf[0], sizeof ebuf);
+    if (library != nullptr) {
+      return library;
+    }
+    if (vm_exit_on_error) {
+      vm_exit(agent, " on the library path, with error: ", missing_module_error_msg);
+    }
+  }
+  return library;
+}
+
+// For absolute and relative paths.
+static void* load_library(JvmtiAgent* agent, const char* on_symbols[], size_t num_symbol_entries, bool vm_exit_on_error) {
+  return agent->is_absolute_path() ? load_agent_from_absolute_path(agent, vm_exit_on_error) :
+                                     load_agent_from_relative_path(agent, vm_exit_on_error);
+}
+
+// Type for the Agent_OnLoad and JVM_OnLoad entry points.
+extern "C" {
+  typedef jint(JNICALL* OnLoadEntry_t)(JavaVM*, char*, void*);
+}
+
+// Find the OnLoad entry point for -agentlib:  -agentpath:   -Xrun agents.
+// num_symbol_entries must be passed-in since only the caller knows the number of symbols in the array.
+static OnLoadEntry_t lookup_On_Load_entry_point(JvmtiAgent* agent, const char* on_load_symbols[], size_t num_symbol_entries) {
+  assert(agent != nullptr, "invariant");
+  if (!agent->is_loaded()) {
+    if (!load_agent_from_executable(agent, on_load_symbols, num_symbol_entries)) {
+      void* const library = load_library(agent, on_load_symbols, num_symbol_entries, /* vm exit on error */ true);
+      assert(library != nullptr, "invariant");
+      agent->set_os_lib(library);
+      agent->set_loaded();
+    }
+  }
+  assert(agent->is_loaded(), "invariant");
+  // Find the OnLoad function.
+  return CAST_TO_FN_PTR(OnLoadEntry_t, os::find_agent_function(agent, false, on_load_symbols, num_symbol_entries));
+}
+
+static OnLoadEntry_t lookup_JVM_OnLoad_entry_point(JvmtiAgent* lib) {
+  const char* on_load_symbols[] = JVM_ONLOAD_SYMBOLS;
+  return lookup_On_Load_entry_point(lib, on_load_symbols, sizeof(on_load_symbols) / sizeof(char*));
+}
+
+static OnLoadEntry_t lookup_Agent_OnLoad_entry_point(JvmtiAgent* agent) {
+  const char* on_load_symbols[] = AGENT_ONLOAD_SYMBOLS;
+  return lookup_On_Load_entry_point(agent, on_load_symbols, sizeof(on_load_symbols) / sizeof(char*));
+}
+
+void JvmtiAgent::convert_xrun_agent() {
+  assert(is_xrun(), "invariant");
+  assert(!is_loaded(), "invariant");
+  assert(JvmtiEnvBase::get_phase() == JVMTI_PHASE_PRIMORDIAL, "invalid init sequence");
+  OnLoadEntry_t on_load_entry = lookup_JVM_OnLoad_entry_point(this);
+  // If there is an JVM_OnLoad function it will get called later,
+  // otherwise see if there is an Agent_OnLoad.
+  if (on_load_entry == nullptr) {
+    on_load_entry = lookup_Agent_OnLoad_entry_point(this);
+    if (on_load_entry == nullptr) {
+      vm_exit_during_initialization("Could not find JVM_OnLoad or Agent_OnLoad function in the library", name());
+    }
+    _xrun = false; // converted
```

Comment:
Just questions to understand it better.
Neither `JVM_Onload` nor `Agent_Onload` entry points are stored after these lookups. It means that in order to be called later (as the comment at line 350 says) they have to be looked up again.
Is it right? Was it the same originally?

Reply:
The entry points are not saved and so have to be looked up again. It was the same originally.

That is why there is a check and branch on agent->is_loaded().

Reply:
Thank you. I'm okay with it.